### PR TITLE
Add storage format version to RocksDBStorage

### PR DESCRIPTION
**Pull request by tillrohrmann**
This commit introduces the notion of a storage format version for
RocksDBStorage. The storage format version is stored in the default
column family under "internal_storage_format_version" and represents
the storage format of all state. When opening RocksDBStorage, it
will check whether the db is empty. If the db is empty, it will write
the current storage format version. If not, then it will check
whether the stored storage format version is compatible with the
current storage format version.

This fixes https://github.com/restatedev/restate/issues/1087.

### CODE REVIEW

**Comment by tillrohrmann** [comment]
I think you are right that storing the version information outside of RocksDB allows to open it differently depending on the version. This is something that I didn't think about. Will change it as I've done it in #1093.

**Comment by AhmedSoliman** [comment]
I'm okay with a file or within rocksdb. The good thing about storing it in rocksdb is that it's taken along when we take database backups/snapshots.

File: crates/storage-rocksdb/src/lib.rs [hunk 1]
```diff
@@ -64,6 +65,16 @@ const FSM_TABLE_NAME: &str = "fsm";
 const TIMERS_TABLE_NAME: &str = "timers";
 const JOURNAL_TABLE_NAME: &str = "journal";
 
+type StorageFormatVersionType = u32;
+
+/// Version of the used storage format. Whenever you introduce a breaking change, this version needs
+/// to be incremented in order to distinguish the new from the old versions. Ideally, you also add
+/// a migration path from the old to the new format.
+const STORAGE_FORMAT_VERSION: StorageFormatVersionType = 0;
```
Comment (AhmedSoliman) [inline crates/storage-rocksdb/src/lib.rs:73]: Tip: Starting with version 1 means that we no hidden default initialization will cause us to believe that we are on the correct version.

**Review by AhmedSoliman — APPROVED** [review]

**Comment by AhmedSoliman** [comment]
> This will help us to decide _how_ to open the database across versions (configuration changes, different CFs, etc’) And it would make reporting issues a bit easier, simply open the text file. If you have a strong preference to actually writing that into Rocksdb, let me know, perhaps I’m overlooking something, otherwise I’d vote for a simple text file.

@tillrohrmann @igalshilman I'm not sure this is a problem we should be worried too much about. The storage format version should live as close as possible to the data (in some possible future, even encoded in the serialized values on per-key basis). 

If opening the column families depend on the version, we would open the default column family first, read the version, then open the rest of column families based on that information, having it in the default column family means makes this easy.

File: crates/storage-rocksdb/src/lib.rs [hunk 2]
```diff
@@ -64,6 +65,16 @@ const FSM_TABLE_NAME: &str = "fsm";
 const TIMERS_TABLE_NAME: &str = "timers";
 const JOURNAL_TABLE_NAME: &str = "journal";
 
+type StorageFormatVersionType = u32;
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/storage-rocksdb/src/lib.rs:68]: True. Will remove it.

File: crates/storage-rocksdb/src/lib.rs (hunk 1 above)
Comment (tillrohrmann) [reply crates/storage-rocksdb/src/lib.rs:73]: This is true. Will change it to `1`.

File: crates/storage-rocksdb/src/lib.rs [hunk 3]
```diff
@@ -423,6 +446,62 @@ impl RocksDBStorage {
...
+                    .expect("The storage format version needs to be an u32.");
+                u32::from_be_bytes(bytes)
+            })
+            .unwrap_or_default();
+
+        if version != STORAGE_FORMAT_VERSION {
+            Err(BuildError::IncompatibleStorageFormat(version))
+        } else {
+            Ok(())
+        }
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply crates/storage-rocksdb/src/lib.rs:502]: Yes indeed.

**Comment by tillrohrmann** [comment]
@AhmedSoliman thanks for the review. What I do like about the version information being stored in RocksDB is that it is then somewhat self-contained as you've said. 

I don't have a strong opinion where the version information should be stored as long as we have it to make migrations possible.

**Pull request merged**